VQ-diffusion

williamberman opened the pull request:
### VQ-diffusion

original issue: https://github.com/huggingface/diffusers/issues/319

So far, this PR only focusses on the vq-diffusion model for the ITHQ dataset. It is possible that the models for the other datasets require additional changes.

The commits are broken out by the model component they port. See the individual commit messages for more in depth write ups on changes

##### Example usage

See linked notebook for how to produce pretrained model for diffusers.

```python
from diffusers.pipelines import VQDiffusionPipeline
import torch
from PIL import Image

OUTPUT_FILE = "<File to save image grid>"

torch.manual_seed(0)

def image_grid(imgs, rows, cols):
    assert len(imgs) == rows*cols
    w, h = imgs[0].size
    grid = Image.new('RGB', size=(cols*w, rows*h))
    for i, img in enumerate(imgs):
        grid.paste(img, box=(i%cols*w, i//cols*h))
    return grid

device = 'cuda'

pipeline = VQDiffusionPipeline.from_pretrained("microsoft/vq-diffusion-ithq").to(device)

images = pipeline("teddy bear playing in the pool", truncation_rate=0.86, num_images_per_prompt=4)

image_grid(images.images, 2, 2).save(OUTPUT_FILE)
```

![out](https://user-images.githubusercontent.com/6505431/197618732-dc6b40d7-d9e0-4499-9540-ce4ab782ae81.png)


##### Progress

- [x] Port VQVAE
- [x] Port transformer 
- [x] VQDiffusionTransformer documentation
- [x] DalleMaskImageEmbedding documentation and clean up
- [x] Port text embeddings (CLIP)
- [x] Port scheduler
- [x] Scheduler documentation
- [x] Use torch.Generator
- [x] VQDiffusion Pipeline docs
- [x] General clean up
- [x] Updating notebook
- [x] Tests?


##### Comparing against original model

I put together a python notebook that compares the currently ported components against eachother. It can be run on google  colab and will handle installing dependencies and model weights. See its README for more details.

https://github.com/williamberman/vq-diffusion-notebook

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_658). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented:
Wooow - this looks amazing! @patil-suraj mind taking a look here?

williamberman commented:
> Wooow - this looks amazing! @patil-suraj mind taking a look here?

Thanks! Would love to know if the minimum unit to merge is the completed pipeline or if smaller chunks like this are acceptable to merge.

345ishaan commented on scripts/convert_vq_diffusion_to_diffusers.py:
```diff
@@ -0,0 +1,508 @@
+
+    original_config = OmegaConf.create(original_config).model
+
+    checkpoint = torch.load(args.checkpoint_path)['model']
```
nit: by default, torch loads model from CUDA. add a flag to specify device-type?

williamberman replied:
sg! added on the transformer branch

patrickvonplaten commented:
Hey @williamberman and @345ishaan,

Sorry for the delay - I have more time going forward and think we can merge this model by next week!

It's a super nice PR. Extremely easy to understand and to follow - thanks a bunch! 

I've fiddled a bit into the PR to make it a bit more light-weight :-) We don't really need a new attention layer as Conv2D layers when used for attention are just like linear layers for which an attention class already exists. 

That's why I removed the `Conv2DAttention` class and changed the conversion script slightly so that your script still works as expected. I'm getting visually identical reconstruction images, so I think we're good with the linear attention layer that was already implemented (could you maybe double check?).

Now I think in a next step we can implement the U-Net, no? Do you need help here?

williamberman commented:
> Hey @williamberman and @345ishaan,
> 
> Sorry for the delay - I have more time going forward and think we can merge this model by next week!
> 
> It's a super nice PR. Extremely easy to understand and to follow - thanks a bunch!
> 
> I've fiddled a bit into the PR to make it a bit more light-weight :-) We don't really need a new attention layer as Conv2D layers when used for attention are just like linear layers for which an attention class already exists.
> 
> That's why I removed the `Conv2DAttention` class and changed the conversion script slightly so that your script still works as expected. I'm getting visually identical reconstruction images, so I think we're good with the linear attention layer that was already implemented (could you maybe double check?).
> 
> Now I think in a next step we can implement the **U-Net** and scheduler for the forward pass, no? Do you need help/guidance here?

Thanks! Pinged in discord as well but the model has a transformer for the reverse diffusion process instead of a unet. I have the transformer ported on another branch. I think the open question is would you prefer that on this PR or to merge this PR first and then merge the transformer on a separate PR?

patrickvonplaten commented:
> > Hey @williamberman and @345ishaan,
> > Sorry for the delay - I have more time going forward and think we can merge this model by next week!
> > It's a super nice PR. Extremely easy to understand and to follow - thanks a bunch!
> > I've fiddled a bit into the PR to make it a bit more light-weight :-) We don't really need a new attention layer as Conv2D layers when used for attention are just like linear layers for which an attention class already exists.
> > That's why I removed the `Conv2DAttention` class and changed the conversion script slightly so that your script still works as expected. I'm getting visually identical reconstruction images, so I think we're good with the linear attention layer that was already implemented (could you maybe double check?).
> > Now I think in a next step we can implement the **U-Net** and scheduler for the forward pass, no? Do you need help/guidance here?
> 
> Thanks! Pinged in discord as well but the model has a transformer (just the encoder iirc) for the reverse diffusion process instead of a unet. I have the transformer ported on another branch. I think the open question is would you prefer that on this PR or to merge this PR first and then merge the transformer on a separate PR?

Hey @williamberman, 

Great to hear that you already have it on a branch. Could you maybe add it directly to this PR? Maybe in a next step we could verify that a forward pass through the transformer (replacement of the unet) gives identical results to the official implementation. If that works, we can integrate the scheduler and then in a last step make a whole denoising process work :-) 

Overall, everything should ideally be in this PR. Since VQ-diffusion will be one of our first community pipeline contributions, there are lots of new things in this PR and I'm more than happy to help you with it (don't hesitate to ping me :-))

williamberman commented:
> Great to hear that you already have it on a branch. Could you maybe add it directly to this PR? Maybe in a next step we could verify that a forward pass through the transformer (replacement of the unet) gives identical results to the official implementation. If that works, we can integrate the scheduler and then in a last step make a whole denoising process work :-)
> 
> Overall, everything should ideally be in this PR. Since VQ-diffusion will be one of our first community pipeline contributions, there are lots of new things in this PR and I'm more than happy to help you with it (don't hesitate to ping me :-))

SG! 

Follow up:
1. Merge transformer into this branch.
2. Add script I've been using to test transformer to readme.
3. Merge CLIP in pipeline/convert script for text embeddings into this branch
4. Add script for testing CLIP to readme
5. Add initial skeleton for scheduler/pipeline to this branch (I also have this on branch with transformer)

patrickvonplaten commented:
> > Great to hear that you already have it on a branch. Could you maybe add it directly to this PR? Maybe in a next step we could verify that a forward pass through the transformer (replacement of the unet) gives identical results to the official implementation. If that works, we can integrate the scheduler and then in a last step make a whole denoising process work :-)
> > Overall, everything should ideally be in this PR. Since VQ-diffusion will be one of our first community pipeline contributions, there are lots of new things in this PR and I'm more than happy to help you with it (don't hesitate to ping me :-))
> 
> SG!
> 
> Follow up:
> 
> 1. Merge transformer into this branch.
> 2. Add script I've been using to test transformer to pr description.
> 3. Merge CLIP in pipeline/convert script for text embeddings into this branch
> 4. Add script for testing CLIP to pr description
> 5. Add initial skeleton for scheduler/pipeline to this branch (I also have this on branch with transformer)

This sounds like a great plan!

williamberman commented:
PR description is updated to reflect progress on merging full model. Notebook which compares outputs from autoencoder, transformer, and text embedder is here and linked in PR description https://github.com/williamberman/vq-diffusion-notebook. Once the scheduler is complete, will also add it to the notebook :)

patrickvonplaten commented:
@williamberman let me know if you'd like me to review already now or better once the scheduler is integrated as well :-)

patrickvonplaten commented:
Great progress!

williamberman commented:
> @williamberman let me know if you'd like me to review already now or better once the scheduler is integrated as well :-)

Let’s wait until the scheduler is integrated. I cleaned some non scheduler components up while working on it that I’d like to add to this branch first :)

williamberman commented:
The working scheduler is on this branch now. Going to do some more cleaning and docs before requesting formal review :)

williamberman commented:
@patrickvonplaten Ok, I think this is ready for reviews! I updated the notebook to compare latents during the denoising process and to compare the final images. The current model is definitely not pixel for pixel the exact same as the original model, but the output images are very similar (see example in notebook). I’m not sure what the diffusers standard for replicating models is here but happy to do some more digging on what could be causing the differences — i.e. it looks like the vendored clip in the original repo produces slightly different embeddings than the clip from transformers we use here. 

Going to take a look at what diffusers requires in terms of tests and add what’s necessary but don’t think that should be a blocker from starting some reviews :)

patrickvonplaten commented:
Very cool! Will try to review this week :-)

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -225,10 +225,18 @@ class CrossAttention(nn.Module):
+        heads: int = 8,
+        dim_head: int = 64,
+        dropout: float = 0.0,
+        bias=False,
```
Nice!

patrickvonplaten commented on src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:
```diff
@@ -0,0 +1,250 @@
+from dataclasses import dataclass
```
The pipeline file looks great!

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+        return hidden_states
+
+
+class VQDiffusionFeedForward(nn.Module):
```
That's quite a normal feed-forward class I think - could we maybe try to re-use: https://github.com/huggingface/diffusers/blob/888468dd90b4e353225d7d505fa28d2963e65678/src/diffusers/models/attention.py#L345 (adding some more default values that are set here

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+        return emb
+
+
+class VQDiffusionTransformerBlock(nn.Module):
```
This is quite a normal attention block - could we maybe re-use https://github.com/huggingface/diffusers/blob/888468dd90b4e353225d7d505fa28d2963e65678/src/diffusers/models/attention.py#L168 here?

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+
+
+class AdaLayerNorm(nn.Module):
+    """
```
Could we add this to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention.py ? :-)

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+        return x
+
+
+class GELU2(nn.Module):
```
Could we add this to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention.py ? :-)

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+from .attention import CrossAttention
+
+
+class VQDiffusionTransformer(ModelMixin, ConfigMixin):
```
Could we try to re-use https://github.com/huggingface/diffusers/blob/888468dd90b4e353225d7d505fa28d2963e65678/src/diffusers/models/attention.py#L107 here as much as possible? #784

patrickvonplaten commented on src/diffusers/models/vq_diffusion_attention.py:
```diff
@@ -0,0 +1,246 @@
+from typing import Optional
```
I think we don't really need a new file here as we should be able to pretty much integrate everything into the existing `attention.py`

williamberman replied:
done!

patrickvonplaten commented on src/diffusers/models/unet_blocks.py:
```diff
@@ -109,6 +109,19 @@ def get_down_block(
             resnet_groups=resnet_groups,
             downsample_padding=downsample_padding,
         )
+    elif down_block_type == "AttnDownEncoderBlock2D":
```
great!

patrickvonplaten commented on src/diffusers/schedulers/scheduling_vq_diffusion.py:
```diff
@@ -0,0 +1,484 @@
+from dataclasses import dataclass
```
Super nice scheduler file!

patrickvonplaten reviewed (commented):
This PR is already in a great shape! 

The main final change would be to try to use existing modules from `attention.py` to a maximum (by adapting them and adding more configurable parameters). This is quite difficult especially given that this code is still quite new. If you like, I could go into the PR here and help you a bit :-) 

Apart from this I think we can merge this PR this week (sorry to be so late here!) 

Could you also add the scheduler and pipeline to the docs under:
- https://github.com/huggingface/diffusers/tree/main/docs/source/api/pipelines
- https://github.com/huggingface/diffusers/blob/main/docs/source/api/schedulers.mdx
? 

Super exciting and very nice work. I think we could have this ready for the release on Thursday - happy to help get this merged on Wednesday :-)  (cc @anton-l @pcuenca @patil-suraj )

williamberman commented:
@patrickvonplaten Ok, I merged the changes into attention.py so the model uses SpatialTransformer. I tried to follow the existing coding style as best as possible. Let me know if you want anything changed there :)

I also added tests for the changed models, the scheduler, and the whole pipeline. Also lmk if I missed another part of the codebase I'm supposed to add tests to.

It looks like the failing tests are from a different PR that was merged recently

williamberman commented on tests/pipelines/vq_diffusion/test_vq_diffusion.py:
```diff
@@ -0,0 +1,150 @@
+        assert np.abs(image_from_tuple_slice.flatten() - expected_slice).max() < 1e-2
+
+
+@slow
+@require_torch_gpu
+class VQDiffusionPipelineIntegrationTests(unittest.TestCase):
+    def tearDown(self):
+        # clean up the VRAM after each test
+        super().tearDown()
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @unittest.skip("VQ Diffusion model not saved to hub")
+    def test_vq_diffusion(self):
+        pass
```
I'm not sure what the process is of uploading this model to the hub

patrickvonplaten replied:
Will take care of it :-)

patrickvonplaten commented on tests/pipelines/vq_diffusion/test_vq_diffusion.py:
```diff
@@ -0,0 +1,174 @@
+        torch.cuda.empty_cache()
+
+    def test_vq_diffusion(self):
+        expected_image = load_image(
```
added a test here @williamberman - pipeline works like a charm and fits nicely with the existing API :-)

williamberman replied:
Great, exciting!

patrickvonplaten commented:
Hey @williamberman,

You've really done an amazing job here! Very impressed by how you were able to add such a new complex model into the existing API! 

The conversion script and your notebook is very easy to follow :-)

I've uploaded the ithq model now to the microsoft org here: https://huggingface.co/microsoft/vq-diffusion-ithq and added a slow test that makes sure the model works as expected. 
Besides that I've done some minor naming changes.

The failing tests are unrelated and we can merge this PR more or less already as is. If ok for you, I would do some final renaming changes tomorrow morning (Paris time) to make it fit a bit better with existing configuration names (will have to sync with @patil-suraj @anton-l and @pcuenca ) and then we can merge this to be in the next release IMO.

@patil-suraj @pcuenca @anton-l could you maybe already review this PR? IMO, besides some re-naming it's ready! I've also made sure that all existing slow tests are passing! 


@williamberman if you're interesting we could do the following follow-up projects to promote this model a bit more:
- Write a short blog post about this model and put in on https://huggingface.co/blog (If you want you could open a blog here: https://github.com/huggingface/blog) - I think the community might be really interested in finding out the difference between latent diffusion models and this vq-diffusion model :-) 
- Make the model card a bit nicer: https://huggingface.co/microsoft/vq-diffusion-ithq (if you want you could open a PR on the repo)
- Send me an email to patrick@huggingface.co and I could connect you to the authors of vq-diffusion model (we could sync with them a bit on promoting this integration on Twitter/Linked-In if you want :-) ) 

Obviously no need to do any of the above points if you don't want or are too busy! Regarding this PR, I think we can merge it tomorrow morning!

Really great job here :rocket:

williamberman commented:
> Hey @williamberman,
> 
> You've really done an amazing job here! Very impressed by how you were able to add such a new complex model into the existing API!
> 
> The conversion script and your notebook is very easy to follow :-)
> 
> I've uploaded the ithq model now to the microsoft org here: https://huggingface.co/microsoft/vq-diffusion-ithq and added a slow test that makes sure the model works as expected. Besides that I've done some minor naming changes.
> 
> The failing tests are unrelated and we can merge this PR more or less already as is. If ok for you, I would do some final renaming changes tomorrow morning (Paris time) to make it fit a bit better with existing configuration names (will have to sync with @patil-suraj @anton-l and @pcuenca ) and then we can merge this to be in the next release IMO.
> 
> @patil-suraj @pcuenca @anton-l could you maybe already review this PR? IMO, besides some re-naming it's ready! I've also made sure that all existing slow tests are passing!
> 
> @williamberman if you're interesting we could do the following follow-up projects to promote this model a bit more:
> 
> * Write a short blog post about this model and put in on https://huggingface.co/blog (If you want you could open a blog here: https://github.com/huggingface/blog) - I think the community might be really interested in finding out the difference between latent diffusion models and this vq-diffusion model :-)
> * Make the model card a bit nicer: https://huggingface.co/microsoft/vq-diffusion-ithq (if you want you could open a PR on the repo)
> * Send me an email to [patrick@huggingface.co](mailto:patrick@huggingface.co) and I could connect you to the authors of vq-diffusion model (we could sync with them a bit on promoting this integration on Twitter/Linked-In if you want :-) )
> 
> Obviously no need to do any of the above points if you don't want or are too busy! Regarding this PR, I think we can merge it tomorrow morning!
> 
> Really great job here 🚀

Awesome all sound good!

I think a blog post sounds great, sent you an email :)

345ishaan commented:
@williamberman I am wondering did you try training it or just verified in infer mode?

williamberman commented:
> @williamberman I am wondering did you try training it or just verified in infer mode?

Just inference using the weights microsoft published. Training would have been a good amount more work 😅

anton-l commented on src/diffusers/schedulers/scheduling_vq_diffusion.py:
```diff
@@ -0,0 +1,480 @@
+from dataclasses import dataclass
```
```suggestion
# Copyright 2022 Microsoft and The HuggingFace Team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from dataclasses import dataclass
```

anton-l commented on src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:
```diff
@@ -0,0 +1,240 @@
+from typing import Callable, List, Optional, Tuple, Union
```
```suggestion
# Copyright 2022 Microsoft and The HuggingFace Team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Callable, List, Optional, Tuple, Union
```

anton-l reviewed (commented):
Awesome integration @williamberman, thank you for working on it!

pcuenca commented on src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:
```diff
@@ -0,0 +1,240 @@
+                The output format of the generated image. Choose between
+                [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
+            return_dict (`bool`, *optional*, defaults to `True`):
+                Whether or not to return a [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] instead of a
```
```suggestion
                Whether or not to return a [`~pipeline_utils.ImagePipelineOutput`] instead of a
```

Maybe should be exported by `pipelines.__init__`? /cc @anton-l

pcuenca commented on src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:
```diff
@@ -0,0 +1,240 @@
+                called at every step.
+
+        Returns:
+            [`~pipeline_utils.ImagePipelineOutput`] or `tuple`: [`~pipelines.utils.ImagePipelineOutput`] if
```
```suggestion
            [`~pipeline_utils.ImagePipelineOutput`] or `tuple`: [`~ pipeline_utils.ImagePipelineOutput `] if
```

pcuenca commented on src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py:
```diff
@@ -0,0 +1,240 @@
+
+        # NOTE: This additional step of normalizing the text embeddings is from VQ-Diffusion.
+        # While CLIP does normalize the pooled output of the text transformer when combining
+        # the image and text embeddings, CLIP does not directly normalize the last hidden state.
```
👍

pcuenca reviewed (approved):
Amazing work!
